cli/fossilize_replay_linux.hpp

```diff
@@ -476,6 +479,209 @@ bool ProcessProgress::start_child_process()
+			long dirty_pages_kb = strtol(buffer + 6, nullptr, 10);
+			if (errno == 0)
+			{
+				state.dirty_pages_mib = dirty_pages_kb / 1024;
+				if (Global::control_block)
+					Global::control_block->dirty_pages_mib.store(state.dirty_pages_mib, std::memory_order_relaxed);
+				got_dirty = true;
+			}
+		}
+	}
+
+	fclose(file);
+	return got_dirty;
+}
+
+static bool get_stall_info(const char *path, uint64_t &total_us)
+{
+	FILE *file = fopen(path, "r");
+	if (!file)
+		return false;
+
+	bool ret = false;
+	char buffer[1024];
+	if (fgets(buffer, sizeof(buffer), file))
+	{
+		if (strncmp(buffer, "some ", 5) == 0)
+		{
+			const char *total = strstr(buffer, "total=");
+			if (total)
+			{
+				errno = 0;
+				total_us = strtoull(total + 6, nullptr, 10);
+				ret = errno == 0;
+			}
+		}
+	}
+
+	fclose(file);
+	return ret;
+}
+
+static bool poll_stall_information(StallState &state)
+{
+	if (!get_dirty_page_info(state))
+		return false;
+
+	if (!get_stall_info("/proc/pressure/cpu", state.cpu_stalled_us))
+		return false;
+	if (!get_stall_info("/proc/pressure/memory", state.memory_stalled_us))
+		return false;
+	if (!get_stall_info("/proc/pressure/io", state.io_stalled_us))
+		return false;
+
+	timespec ts = {};
+	if (clock_gettime(CLOCK_MONOTONIC, &ts) != 0)
+		return false;
+	state.timestamp_ns = ts.tv_sec * 1000000000ll + ts.tv_nsec;
+
+	return true;
+}
+
+static void manage_thrashing_behavior(std::vector<ProcessProgress> &child_processes,
+                                      StallState &old_state, const StallState &new_state)
+{
+	// It is quite easy to overload the system with IO if we're not careful, so check for this.
+
+	float delta_us = 1e-3f * float(new_state.timestamp_ns - old_state.timestamp_ns);
+	float cpu_stall_us = new_state.cpu_stalled_us - old_state.cpu_stalled_us;
+	float io_stall_us = new_state.io_stalled_us - old_state.io_stalled_us;
+	float mem_stall_us = new_state.memory_stalled_us - old_state.memory_stalled_us;
+	int32_t delta_dirty_pages_mib = int32_t(new_state.dirty_pages_mib) - int32_t(old_state.dirty_pages_mib);
+
+	float cpu_stall_ratio = cpu_stall_us / delta_us;
+	float io_stall_ratio = io_stall_us / delta_us;
+	float mem_stall_ratio = mem_stall_us / delta_us;
+	cpu_stall_ratio = std::min(1.0f, cpu_stall_ratio);
+	io_stall_ratio = std::min(1.0f, io_stall_ratio);
+	mem_stall_ratio = std::min(1.0f, mem_stall_ratio);
+
+	unsigned target_running_processes = Global::running_processes;
+
+	// Some heuristics here.
+	bool go_down = false;
+	bool go_up = false;
+
+	if (delta_dirty_pages_mib > 50)
+	{
+		// If we (or someone else) created more than 50 MiB of dirty data in this tick,
+		// we should lower number of processes.
+		go_down = true;
+	}
+	else if (delta_dirty_pages_mib < -10)
+	{
+		// Dirty pages is going down, can increase CPU load now.
+		go_up = true;
+	}
+
+	// There are other processes that would like some CPU time.
+	if (cpu_stall_ratio > 0.5f)
+		go_down = true;
+	else if (cpu_stall_ratio < 0.1f)
+		go_up = true;
```

Comment:
Should we skip this check if fossilize runs in foreground mode? (that is when starting a game and Steam shows the progress dialog window for fossilize)

Reply:
BTW: After enabling PSI on my system, I'm pretty sure fossilize should not look at CPU pressure. In my system:
```
# cat /proc/pressure/cpu
some avg10=100.00 avg60=99.99 avg300=100.00 total=45537131575
```

This is because I have a very low priority daemon running that constantly uses at least one core.

So I suggest skipping this CPU pressure check completely and relying on the CPU scheduler instead (which we already switched to batch, with proper niceness of the process group). We could still use the `go_up` path of this check to add more threads when the CPU seems idle enough, tho.

Reply:
However, it looks like `total` is not increasing at a 100% rate although `avg` suggests this... I'll look into that more deeply in the weekend.

Reply:
Here's a graph of CPU pressure vs. CPU utilization vs. loadavg from one of our systems:

![image](https://user-images.githubusercontent.com/17684/103879937-6140fa00-50d0-11eb-995b-68aaa6a88ea1.png)
![image](https://user-images.githubusercontent.com/17684/103880066-8d5c7b00-50d0-11eb-898b-684ba973177f.png)
![image](https://user-images.githubusercontent.com/17684/103880096-95b4b600-50d0-11eb-839d-8c4e0ef4f2bd.png)

loadavg and CPU pressure seem to correlate very well.

This is IO pressure and memory pressure which also seem to correlate:

![image](https://user-images.githubusercontent.com/17684/103880294-d7ddf780-50d0-11eb-92b4-4eb2c8c90843.png)
![image](https://user-images.githubusercontent.com/17684/103880321-df050580-50d0-11eb-8dee-d8598595decd.png)

Reply:
As we can see in the graphs, a stall on IO is almost always a full stall, so we should prioritize on keeping IO stall under control. Memory stall is less of a problem, it seems.

Reply:
> I think you can write a desired threshold to the file, then use `select/poll/epoll()`. That may be more accurate than manually calculating the values from diffs.

I tried this, but it requires root on my system since the /proc/pressure files are read-only for normal users, and we have to write to it to become a monitor. I think it's fine to do it like this. There is a timerfd that fires at regular intervals in the master process which does periodic polling anyways. We need to know the ratio, not just if a threshold was hit anyways.

Reply:
I agree on CPU stall factors being kinda meaningless.

I changed it to just consider IO related factors.

Reply:
IO and memory may no longer correlate when the system is under memory pressure... OTOH, if swap is involved, IO will start to correlate again. In theory, this means we can get away with watching IO pressure only. In practice, I didn't test it and probably won't because my system was never affected in the way that other users reported (OOM killer kicking in etc).

Reply:
> I tried this, but it requires root on my system

Ah okay... I also agree for the remainder of your answer.

Reply:
> I agree on CPU stall factors being kinda meaningless.

Yeah, I think we should also stop using `nice` when Steam is running fossilize in foreground processing mode. `SCHED_BATCH` is still useful, tho (because better throughput using longer scheduler timeslices).

Reply:
Is it possible for fossilize to detect foreground vs background mode?

Reply:
I'm considering adding a way for the controlling process to dynamically enable/disable PSI usage. Not sure if we should bother with nice. Since we now have a setup to control number of active processes dynamically it should be easy to say something like "only keep N processes active in background, but when going to foreground, light up all cores", something like that.

Reply:
> dynamically enable/disable PSI usage

I think PSI usage should be watched the whole time, no matter if we a limited to all cores or just a reduced number of cores (background mode). The original issue report was about fossilize overwhelming the available system resources - and watching IO pressure is the only active thing we can do to prevent that situation. While the user waits for a foreground fossilize to finish, they still might want to use the system for other tasks. Similarly, watching the dirty cache ratio is still useful but may be relaxed in foreground mode.

BTW: Is it possible to just replace the fossilize binary deployed by Steam with a compiled version from this branch for testing? OTOH, it won't run anyways currently because I'm on NVIDIA. Is there a way to override that behavior?

Reply:
It's possible to replace the fossilize replayer binary I think. However, the external launcher code is linked into Steam I think, so that's harder to change. I don't know much if anything about the Steam side, I just work on this repo.

Reply:
Ok, made it so that the external replayer can send messages to the replayer about options. It can then force number of active processes, enable if IO stalling matters, etc. That should enable Steam to dynamically change between background/foreground type scheduling.

Going beyond this will be extreme overkill, so I'll probably leave it here. Need to test a bit more later.

Reply:
> However, the external launcher code is linked into Steam I think

Ah okay yes, that's problematic then. This probably also means it cannot be distributed as a separate package which could be switched to beta (like Proton).

> That should enable Steam to dynamically change between background/foreground type scheduling.

I think that's more than enough. I just asked if it would be possible. ;-)